### PR TITLE
Added IFilterable and SearchContainer

### PR DESCRIPTION
**Pull request by Jorolf**

### CODE REVIEW

File: osu.Framework/Graphics/Containers/ISearchable.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu-framework/master/LICENCE
+
+using System.Collections.Generic;
+
+namespace osu.Framework.Graphics.Containers
+{
+    public interface ISearchable
+    {
+        /// <summary>
+        /// List of keywords that <see cref="SearchContainer"/> should match
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ISearchable.cs]: `ISearchable` shouldn't know about `SearchContainer`.

The documentation on how to use `SearchContainer` should be on the `SearchContainer` class itself, explaining these two interfaces.

File: osu.Framework/Graphics/Containers/ISearchable.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        /// </summary>
+        string[] Keywords { get; }
+        /// <summary>
+        /// Current matching state
+        /// </summary>
+        bool Matching { set; }
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ISearchable.cs]: `FilteredByParent`?

**Comment by peppy** [comment]
still not 100% sure about recursive logic, but at very least you need to have an example of recursive logic in the testcase.

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+namespace osu.Framework.Graphics.Containers
+{
+    public class SearchContainer : SearchContainer<Drawable>
+    { }
+
+    public class SearchContainer<T> : Container<T>, ISearchableChildren where T : Drawable
+    {
+        private string searchTerm;
+        /// <summary>
+        /// String to match with the <see cref="ISearchable"/>s
```
Comment (Jorolf) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: that was intended

File: osu.Framework/Graphics/Containers/IFilterable.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace osu.Framework.Graphics.Containers
+{
+    public interface IFilterable
+    {
+        /// <summary>
+        /// Array of keywords that it can be filtered with
+        /// </summary>
+        string[] Keywords { get; }
+        /// <summary>
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/IFilterable.cs]: newline

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                match(this);
+            }
+        }
+
+        public IEnumerable<IFilterable> FilterableChildren => Children.OfType<IFilterable>();
+        public string[] Keywords => null;
+        public bool FilteredByParent { get; set; }
+
+
+        private bool match(IFilterable searchable, List<string> parentKeywords = null)
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: `parentKeywords` should probably just be `keywords`.

not even sure if `keywords` is correct as they could be more than one word. maybe use `terms` instead?

File: osu.Framework.VisualTests/Tests/TestCaseSearchContainer.cs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+            new List<string>
+            {
+                "test",
+                "sUbSeCtIoN 2",
+                "€",
+                "èê",
+                "321",
+                "?!"
+            }.ForEach(term => AddStep("Search term: " + term, () => search.SearchTerm = term));
```
Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseSearchContainer.cs]: add assert steps to ensure the number of displaying matches is correct in each case.

File: osu.Framework/Graphics/Containers/IFilterable.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu-framework/master/LICENCE
+
+using System.Collections.Generic;
+
+namespace osu.Framework.Graphics.Containers
+{
+    public interface IFilterable
+    {
+        /// <summary>
+        /// Array of terms that it can be filtered with
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/IFilterable.cs]: An array of relevant terms which match the current object in a filtered scenario.

File: osu.Framework/Graphics/Containers/IFilterable.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        /// </summary>
+        string[] Terms { get; }
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/IFilterable.cs]: FilterTerms

File: osu.Framework/Graphics/Containers/IFilterable.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+        /// <summary>
+        /// Current filtered state, changes whenever it's getting filtered
+        /// </summary>
+        bool FilteredByParent { set; }
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/IFilterable.cs]: `MatchingCurrentFilter`

File: osu.Framework/Graphics/Containers/IFilterable.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+{
+    public interface IFilterable
+    {
+        /// <summary>
+        /// Array of terms that it can be filtered with
+        /// </summary>
+        string[] Terms { get; }
+
+        /// <summary>
+        /// Current filtered state, changes whenever it's getting filtered
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/IFilterable.cs]: Whether the current object is matching (ie. visible) given the current filter criteria of a parent.

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace osu.Framework.Graphics.Containers
+{
+    public class SearchContainer : SearchContainer<Drawable>
+    {
+    }
+
+    public class SearchContainer<T> : Container<T>, IFilterableChildren where T : Drawable
+    {
+        private string searchTerm;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: newline

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+        /// <summary>
+        /// String to filter <see cref="IFilterable"/>
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: `see cref` should be used in a sentence.

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+            childTerms.AddRange(searchable.Terms ?? new string[0]);
+
+            var searchContainer = searchable as IFilterableChildren;
+
+            if (searchContainer != null)
+            {
+                bool matching = false;
+
+                foreach (IFilterable searchableChildren in searchContainer.FilterableChildren)
+                    matching = match(searchableChildren, childTerms) || matching;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: `matching |=`

File: osu.Framework/Graphics/Containers/IFilterable.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        /// </summary>
+        bool FilteredByParent { set; }
+    }
+
+    public interface IFilterableChildren : IFilterable
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/IFilterable.cs]: IHasFilterableChildren

Comment (peppy) [inline osu.Framework/Graphics/Containers/IFilterable.cs]: probably in its own file

File: osu.Framework/Graphics/Containers/IFilterable.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        /// <summary>
+        /// List of children that should be filtered
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/IFilterable.cs]: ...that can be filtered

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace osu.Framework.Graphics.Containers
+{
+    public class SearchContainer : SearchContainer<Drawable>
+    {
+    }
+
+    public class SearchContainer<T> : Container<T>, IFilterableChildren where T : Drawable
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: is there a reason SearchContainer needs to implement `IFilterableChildren`? It feels like it doesn't need to be a thing.

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        public bool FilteredByParent { get; set; }
+
+        private bool match(IFilterable searchable, IEnumerable<string> terms)
+        {
+            var childTerms = new List<string>(terms);
+            childTerms.AddRange(searchable.Terms ?? new string[0]);
+
+            var searchContainer = searchable as IFilterableChildren;
+
+            if (searchContainer != null)
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: don't call this searchContainer.

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+            {
+                bool matching = false;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: i'd probably add a comment here mentioning that we have children we also need to check (and mention that we ourselves should match if any recursive child matches)

File: osu.Framework/Graphics/Containers/SearchContainer.cs (hunk 16 above)
Comment (Jorolf) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: I added that so I could use the `match(IFilterable filterable, IEnumerable<string> terms)` method and don't have to iterate over the children

Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: i'd say iterating over children may be a better solution.

by doing this, you're exposing SearchContainer as a filterable container when it isn't actually correctly handling the interface.

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+            set
+            {
+                searchTerm = value;
+                match(this, new string[0]);
+            }
+        }
+
+        public IEnumerable<IFilterable> FilterableChildren => Children.OfType<IFilterable>();
+        public string[] Terms => null;
+        public bool FilteredByParent { get; set; }
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: it doesn't

File: osu.Framework/Graphics/Containers/SearchContainer.cs [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+            }
+            set
+            {
+                searchTerm = value;
+                match(this, new string[0]);
+            }
+        }
+
+        public IEnumerable<IFilterable> FilterableChildren => Children.OfType<IFilterable>();
+        public string[] Terms => null;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/SearchContainer.cs]: what is null? undefined behaviour (the interface does not specify it can be null)

**Pull request merged**